cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -936,6 +937,101 @@ impl Inst {
+            Inst::mov_r_r(ty.bits() == 64, from_reg, to_reg)
+        } else {
+            let opcode = match ty {
+                types::F32 => SseOpcode::Movss,
+                types::F64 => SseOpcode::Movsd,
+                types::F32X4 => SseOpcode::Movaps,
+                types::F64X2 => SseOpcode::Movapd,
+                _ if ty.is_vector() && ty.bits() == 128 => SseOpcode::Movdqa,
+                _ => unimplemented!("unable to move type: {}", ty),
+            };
+            Inst::xmm_unary_rm_r(opcode, RegMem::reg(from_reg), to_reg)
+        }
+    }
+
+    /// Choose which instruction to use for loading a register value from memory. This currently
+    /// zero extends any loads smaller than 64 bits but should eventually support sign extension
+    /// (TODO).
+    pub(crate) fn load(
+        ty: Type,
+        from_addr: impl Into<SyntheticAmode>,
+        to_reg: Writable<Reg>,
+        srcloc: Option<SourceLoc>,
+    ) -> Inst {
+        if is_movable_in_scalar_reg(ty) {
+            let ext_mode = match ty.bytes() {
+                1 => Some(ExtMode::BQ),
+                2 => Some(ExtMode::WQ),
+                4 => Some(ExtMode::LQ),
+                8 => None,
+                _ => unreachable!("the type should never use a scalar load: {}", ty),
+            };
+            if let Some(ext_mode) = ext_mode {
+                Inst::movzx_rm_r(ext_mode, RegMem::mem(from_addr), to_reg, srcloc)
+            } else {
+                Inst::mov64_m_r(from_addr, to_reg, srcloc)
+            }
+        } else {
+            let opcode = match ty {
+                types::F32 => SseOpcode::Movss,
+                types::F64 => SseOpcode::Movsd,
+                types::F32X4 => SseOpcode::Movups,
+                types::F64X2 => SseOpcode::Movupd,
+                _ if ty.is_vector() && ty.bits() == 128 => SseOpcode::Movdqu,
+                _ => unimplemented!("unable to load type: {}", ty),
+            };
+            Inst::xmm_unary_rm_r(opcode, RegMem::mem(from_addr), to_reg)
+        }
+    }
+
+    /// Choose which instruction to use for storing a register value to memory.
+    pub(crate) fn store(
+        ty: Type,
+        from_reg: Reg,
+        to_addr: impl Into<SyntheticAmode>,
+        srcloc: Option<SourceLoc>,
+    ) -> Inst {
+        if is_movable_in_scalar_reg(ty) {
+            Inst::mov_r_m(ty.bytes() as u8, from_reg, to_addr, srcloc)
+        } else {
+            let opcode = match ty {
+                types::F32 => SseOpcode::Movss,
+                types::F64 => SseOpcode::Movsd,
+                types::F32X4 => SseOpcode::Movups,
+                types::F64X2 => SseOpcode::Movupd,
+                _ if ty.is_vector() && ty.bits() == 128 => SseOpcode::Movdqu,
+                _ => unimplemented!("unable to store type: {}", ty),
+            };
+            Inst::xmm_mov_r_m(opcode, from_reg, to_addr, srcloc)
+        }
+    }
+}
+
+fn is_movable_in_scalar_reg(ty: Type) -> bool {
```

Comment:
A doc comment would be helpful here -- from the name it's not immediately clear what this does (movable to/from a scalar register? movable within a scalar register? ...).

Reply:
Got rid of this and used a comparison on `RegClass` instead.